### PR TITLE
Validate input sample rate for decoding is a valid value as per opus documentation

### LINKED ISSUE #4: The library is making assumptions about the source file format.
**Describe the bug**
As described in https://github.com/vector-im/element-ios/issues/6603, `OGGConverter` and `OGGDecoder` make assumptions about the format of the file, in particular that the file is mono. This can result in incorrect decoding and playback when a user tries to play an opus file that wasn't sent using the voice message functionality.

**To Reproduce**
See the issue in Element iOS

**Expected behavior**
The library should determine the format of the audio decode from the file rather than having this hard-coded.

**Smartphone (please complete the following information):**
iPhone 12 on iOS 15.6.1 running Element 1.8.27

### PR DESCRIPTION
**Pull request by langleyd**
Fixes https://github.com/vector-im/swift-ogg/issues/4

The existing code blindly accepted the sample rate in the opus file header(if it was within a specific range).

However as per the [opus documentation](https://opus-codec.org/docs/opus_api-1.3.1/group__opus__multistream.html#ga3c0e342774174c471e61cedba53755c9:~:text=opus_multistream_decoder_init()):

> Sampling rate to decode at (in Hz). This must be one of 8000, 12000, 16000, 24000, or 48000.

To fix the issue I have:
1. Fixed the incorrectly handled optional that was causing a crash before we could read the status error from the attempt to create the decoder. We now throw `OpusError.badArgument` error if the decoder is nil.
2. Rather than accepting the sample rate in the header, we now choose the closest valid sample rate to that defined in the header(For example in the bug reported the file had an input sample rate of 44100, this now gets mapped to a valid sample rate of 48000).
3. Added a test to verify opus files with invalid sample rates are handled correctly.

### CODE REVIEW

File: Sources/SwiftOgg/OGGDecoder.swift [hunk 1]
```diff
@@ -226,18 +228,29 @@ class OGGDecoder {
...
+        guard let newDecoder = newDecoder else {
+            throw OpusError.badArgument
+        }
+        
+        decoder = newDecoder
         return decoder
     }
+    
+    static func getClosestValidSampleRate(inputRate: Int32) -> Int32 {
+        guard let closest = validSampleRates.enumerated().min(by: { abs($0.1 - inputRate) < abs($1.1 - inputRate)} ) else {
```
Comment (pixlwave) [inline Sources/SwiftOgg/OGGDecoder.swift:249]: TIL about `enumerated().min()` 🤩

File: Tests/EncoderDecoderTests/ConverterTests.swift [hunk 2]
```diff
@@ -20,6 +20,17 @@ import AVFoundation
...
 class ConverterTests: XCTestCase {
 
+    func testInvalidSampleRate() {
+        let src = Bundle.module.url(forResource: "InvalidSampleRate", withExtension: "ogg")!
+        let dest = URL(fileURLWithPath: NSTemporaryDirectory() + "VoiceRecordingWebOut.m4a")
+        do {
+            try OGGConverter.convertOpusOGGToM4aFile(src: src, dest: dest)
+            XCTAssert(try dest.checkResourceIsReachable(), "destination m4a file does not exist")
+        } catch {
+            XCTAssert(false, "Failed to convert from ogg to m4a with error \(error)")
```
Comment (pixlwave) [inline Tests/EncoderDecoderTests/ConverterTests.swift]: Not important, but could this one be an `XCTFail`, I find those a bit quicker to process in my head as "this is the bad path".

**Review by pixlwave — APPROVED** [review]
Looks great to me 😎

Sidenote: Did you try one of the files linked from https://github.com/vector-im/element-ios/issues/6603? I think that's a different issue, but curious to know if this is actually the fix here too.

**Comment by langleyd** [comment]
@pixlwave It' doesn't fix the problem for https://github.com/vector-im/element-ios/issues/6603. I tried to fix that also, but haven't been able to so far.

**Comment by pixlwave** [comment]
Not to worry, fixing the crash is the important one 🙌

**Pull request merged**